README.md
# lttng-analyses

This repository contains various scripts to extract monitoring data and metrics
from LTTng kernel traces.

As opposed to other diagnostic or monitoring solutions, this approach is designed
to allow users to record their system's activity with a low overhead, wait
for a problem to occur and then diagnose its cause offline.

This solution allows the user to target hard to find problems and dig until the
root cause is found.

This README describes the implemented analyses as well as how to use them.

## Requirements
* LTTng >= 2.5
* Babeltrace >= 1.2 (with python bindings built)
* Python >= 3.4

## Install
### Ubuntu (12.04 and 14.04 at least)
```bash
apt-get install -y software-properties-common (or python-software-properties on 12.04)
apt-add-repository -y ppa:lttng/ppa
apt-get update
apt-get -y install lttng-tools babeltrace lttng-modules-dkms python3-babeltrace python3-progressbar
git clone https://github.com/lttng/lttng-analyses.git
cd lttng-analyses
python setup.py install
```

### Other distros
Please refer to the [LTTng documentation](http://lttng.org/download/) to
install LTTng and the [Babeltrace
README](http://git.efficios.com/?p=babeltrace.git;a=blob_plain;f=README;hb=HEAD)
to install ```babeltrace``` with the python bindings. Optionally install the
```progressbar``` python module, and then:

```bash
git clone https://github.com/lttng/lttng-analyses.git
cd lttng-analyses
python setup.py install
```

## Trace creation
Here are the basic commands to create a trace, for more information on the
LTTng setup, please refer to the [LTTng
documentation](http://lttng.org/docs/#doc-tracing-the-linux-kernel)

### Automatic
From the cloned git tree:
```bash
./lttng-analyses-record
```

### Manual
```bash
lttng create
lttng enable-channel -k bla --subbuf-size=4M
lttng enable-event -k sched_switch,block_rq_complete,block_rq_issue,block_bio_remap,block_bio_backmerge,netif_receive_skb,net_dev_xmit,sched_process_fork,sched_process_exec,lttng_statedump_process_state,lttng_statedump_file_descriptor,lttng_statedump_block_device,writeback_pages_written,mm_vmscan_wakeup_kswapd,mm_page_free,mm_page_alloc,block_dirty_buffer,irq_handler_entry,irq_handler_exit,softirq_entry,softirq_exit,softirq_raise -c bla
lttng enable-event -k --syscall -a -c bla
lttng start
..do stuff...
lttng stop
lttng destroy
```

### Remote trace creation
You can also create a trace on a server and send it to a remote host. The
remote host only needs to run `lttng-relayd -d` and be reachable over the network.
The only difference with the above commands is the tracing session's creation:
```bash
lttng create -U net://<remote-host>
```

## Implemented analyses

* CPU usage for the whole system
* CPU usage per-process
* Process CPU migration count
* Memory usage per-process (as seen by the kernel)
* Memory usage system-wide (as seen by the kernel)
* I/O usage (syscalls, disk, network)
* I/O operations log (with latency and usage)
* I/O latency statistics (open, read, write, sync operations)
* I/O latency frequency distribution
* Interrupt handler duration statistics (count, min, max, average stdev)
* Interrupt handler duration top
* Interrupt handler duration log
* Interrupt handler duration frequency distribution
* SoftIRQ handler latency statistics
* Syscalls usage statistics

All of the analyses share the same code architecture making it possible
to filter by timerange, process name, PID, min and max values using the
same command-line options. Also note that reported timestamps can
optionally be expressed in the GMT timezone to allow easy sharing between
teams.

The project's architecture makes it easy to add new analyses or to reuse
the analysis backend in external tools which may then present the results
in their own format (as opposed to text).

## Examples
After having collected your trace, any script contained in this repository
can be used to run an analysis. Read on for some examples!

### I/O
#### I/O latency stats
```bash
$ ./lttng-iolatencystats mytrace/
Timerange: [2015-01-06 10:58:26.140545481, 2015-01-06 10:58:27.229358936]
Syscalls latency statistics (usec):
Type                    Count            Min        Average            Max          Stdev
-----------------------------------------------------------------------------------------
Open                       45          5.562         13.835         77.683         15.263
Read                      109          0.316          5.774         62.569          9.277
Write                     101          0.256          7.060         48.531          8.555
Sync                      207         19.384         40.664        160.188         21.201

Disk latency statistics (usec):
Name                    Count            Min        Average            Max          Stdev
-----------------------------------------------------------------------------------------
dm-0                      108          0.001          0.004          0.007          1.306
```

#### I/O latency frequency distribution
```bash
$ ./lttng-iolatencyfreq mytrace/
Timerange: [2015-01-06 10:58:26.140545481, 2015-01-06 10:58:27.229358936]
Open latency distribution (usec)
###############################################################################
 5.562 ███████████████████████████████████████████████████████████████████  25
 9.168 ██████████                                                            4
12.774 █████████████████████                                                 8
16.380 ████████                                                              3
19.986 █████                                                                 2
23.592                                                                       0
27.198                                                                       0
30.804                                                                       0
34.410 ██                                                                    1
38.016                                                                       0
41.623                                                                       0
45.229                                                                       0
48.835                                                                       0
52.441                                                                       0
56.047                                                                       0
59.653                                                                       0
63.259                                                                       0
66.865                                                                       0
70.471                                                                       0
74.077 █████                                                                 2
```

#### I/O latency top
```bash
$ ./lttng-iolatencytop analysis-20150115-120942/ --limit 3 --minsize 2
Checking the trace for lost events...
Timerange: [2015-01-15 12:18:37.216484041, 2015-01-15 12:18:53.821580313]
Top open syscall latencies (usec)
Begin               End                  Name             Duration (usec)         Size  Proc                     PID      Filename      
[12:18:50.432950815,12:18:50.870648568]  open                  437697.753          N/A  apache2                  31517    /var/lib/php5/sess_0ifir2hangm8ggaljdphl9o5b5 (fd=13)
[12:18:52.946080165,12:18:52.946132278]  open                      52.113          N/A  apache2                  31588    /var/lib/php5/sess_mr9045p1k55vin1h0vg7rhgd63 (fd=13)
[12:18:46.800846035,12:18:46.800874916]  open                      28.881          N/A  apache2                  31591    /var/lib/php5/sess_r7c12pccfvjtas15g3j69u14h0 (fd=13)
[12:18:51.389797604,12:18:51.389824426]  open                      26.822          N/A  apache2                  31520    /var/lib/php5/sess_4sdb1rtjkhb78sabnoj8gpbl00 (fd=13)

Top read syscall latencies (usec)
Begin               End                  Name             Duration (usec)         Size  Proc                     PID      Filename      
[12:18:37.256073107,12:18:37.256555967]  read                     482.860       7.00 B  bash                     10237    unknown (origin not found) (fd=3)
[12:18:52.000209798,12:18:52.000252304]  read                      42.506      1.00 KB  irqbalance               1337     /proc/interrupts (fd=3)
[12:18:37.256559439,12:18:37.256601615]  read                      42.176       5.00 B  bash                     10237    unknown (origin not found) (fd=3)
[12:18:42.000281918,12:18:42.000320016]  read                      38.098      1.00 KB  irqbalance               1337     /proc/interrupts (fd=3)

Top write syscall latencies (usec)
Begin               End                  Name             Duration (usec)         Size  Proc                     PID      Filename      
[12:18:49.913241516,12:18:49.915908862]  write                   2667.346      95.00 B  apache2                  31584    /var/log/apache2/access.log (fd=8)
[12:18:37.472823631,12:18:37.472859836]  writev                    36.205     21.97 KB  apache2                  31544    unknown (origin not found) (fd=12)
[12:18:37.991578372,12:18:37.991612724]  writev                    34.352     21.97 KB  apache2                  31589    unknown (origin not found) (fd=12)
[12:18:39.547778549,12:18:39.547812515]  writev                    33.966     21.97 KB  apache2                  31584    unknown (origin not found) (fd=12)

Top sync syscall latencies (usec)
Begin               End                  Name             Duration (usec)         Size  Proc                     PID      Filename      
[12:18:50.162776739,12:18:51.157522361]  sync                  994745.622          N/A  sync                     22791    None (fd=None)
[12:18:37.227867532,12:18:37.232289687]  sync_file_range         4422.155          N/A  lttng-consumerd          19964    /home/julien/lttng-traces/analysis-20150115-120942/kernel/metadata (fd=32)
[12:18:37.238076585,12:18:37.239012027]  sync_file_range          935.442          N/A  lttng-consumerd          19964    /home/julien/lttng-traces/analysis-20150115-120942/kernel/metadata (fd=32)
[12:18:37.220974711,12:18:37.221647124]  sync_file_range          672.413          N/A  lttng-consumerd          19964    /home/julien/lttng-traces/analysis-20150115-120942/kernel/metadata (fd=32)
```

#### I/O operations log
```bash
$ ./lttng-iolog mytrace/
[10:58:26.221618530,10:58:26.221620659]  write                      2.129       8.00 B  /usr/bin/x-term          11793    anon_inode:[eventfd] (fd=5)
[10:58:26.221623609,10:58:26.221628055]  read                       4.446      50.00 B  /usr/bin/x-term          11793    /dev/ptmx (fd=24)
[10:58:26.221638929,10:58:26.221640008]  write                      1.079       8.00 B  /usr/bin/x-term          11793    anon_inode:[eventfd] (fd=5)
[10:58:26.221676232,10:58:26.221677385]  read                       1.153       8.00 B  /usr/bin/x-term          11793    anon_inode:[eventfd] (fd=5)
[10:58:26.223401804,10:58:26.223411683]  open                       9.879          N/A  sleep                    12420    /etc/ld.so.cache (fd=3)
[10:58:26.223448060,10:58:26.223455577]  open                       7.517          N/A  sleep                    12420    /lib/x86_64-linux-gnu/libc.so.6 (fd=3)
[10:58:26.223456522,10:58:26.223458898]  read                       2.376     832.00 B  sleep                    12420    /lib/x86_64-linux-gnu/libc.so.6 (fd=3)
[10:58:26.223918068,10:58:26.223929316]  open                      11.248          N/A  sleep                    12420     (fd=3)
[10:58:26.231881565,10:58:26.231895970]  writev                    14.405      16.00 B  /usr/bin/x-term          11793    socket:[45650] (fd=4)
[10:58:26.231979636,10:58:26.231988446]  recvmsg                    8.810      16.00 B  Xorg                     1827     socket:[47480] (fd=38)
```


#### I/O usage top
```bash
$ ./lttng-iousagetop traces/pgread-writes
Timerange: [2014-10-07 16:36:00.733214969, 2014-10-07 16:36:18.804584183]
Per-process I/O Read
###############################################################################
██████████████████████████████████████████████████    16.00 MB lttng-consumerd (2619)         0 B  file   4.00 B  net  16.00 MB unknown
█████                                                  1.72 MB lttng-consumerd (2619)         0 B  file      0 B  net   1.72 MB unknown
█                                                    398.13 KB postgres (4219)           121.05 KB file 277.07 KB net   8.00 B  unknown
                                                     256.09 KB postgres (1348)                0 B  file 255.97 KB net 117.00 B  unknown
                                                     204.81 KB postgres (4218)           204.81 KB file      0 B  net      0 B  unknown
                                                     123.77 KB postgres (4220)           117.50 KB file   6.26 KB net   8.00 B  unknown
Per-process I/O Write
###############################################################################
██████████████████████████████████████████████████    16.00 MB lttng-consumerd (2619)         0 B  file   8.00 MB net   8.00 MB unknown
██████                                                 2.20 MB postgres (4219)             2.00 MB file 202.23 KB net      0 B  unknown
█████                                                  1.73 MB lttng-consumerd (2619)         0 B  file 887.73 KB net 882.58 KB unknown
██                                                   726.33 KB postgres (1165)             8.00 KB file   6.33 KB net 712.00 KB unknown
                                                     158.69 KB postgres (1168)           158.69 KB file      0 B  net      0 B  unknown
                                                      80.66 KB postgres (1348)                0 B  file  80.66 KB net      0 B  unknown
Files Read
###############################################################################
██████████████████████████████████████████████████     8.00 MB anon_inode:[lttng_stream] (lttng-consumerd) 'fd 32 in lttng-consumerd (2619)'
█████                                                834.41 KB base/16384/pg_internal.init 'fd 7 in postgres (4219)', 'fd 7 in postgres (4220)', 'fd 7 in postgres (4221)', 'fd 7 in postgres (4222)', 'fd 7 in postgres (4223)', 'fd 7 in postgres (4224)', 'fd 7 in postgres (4225)', 'fd 7 in postgres (4226)'
█                                                    256.09 KB socket:[8893] (postgres) 'fd 9 in postgres (1348)'
█                                                    174.69 KB pg_stat_tmp/pgstat.stat 'fd 9 in postgres (4218)', 'fd 9 in postgres (1167)'
                                                     109.48 KB global/pg_internal.init 'fd 7 in postgres (4218)', 'fd 7 in postgres (4219)', 'fd 7 in postgres (4220)', 'fd 7 in postgres (4221)', 'fd 7 in postgres (4222)', 'fd 7 in postgres (4223)', 'fd 7 in postgres (4224)', 'fd 7 in postgres (4225)', 'fd 7 in postgres (4226)'
                                                     104.30 KB base/11951/pg_internal.init 'fd 7 in postgres (4218)'

ChangeLog
2016-07-25 LTTng analyses 0.5.4
	* fix: namespace softirq events
	* Fix: schedfreq: append freq result tables in MI mode
	* Fix argument order passed to parse_trace_collection_time_range.
	* Fix: tests trace utils with UTC+1400 TZ

2016-06-10 LTTng analyses 0.5.3
	* Add travis CI support
	* Doc: Add usage and debug to readme
	* Fix: pep8 fixes
	* Fix: tests when shell locale is not UTF-8 again

2016-06-07 LTTng analyses 0.5.2
	* Fix: tests when shell locale is not UTF-8
	* Handle events without fields
	* Sanitize event names when using them as variable/function name
	* Fix: command.py: pass params in order to parse_trace_collection_date()

2016-05-21 LTTng analyses 0.5.1
	* Fix missing import

2016-05-27 LTTng analyses 0.5.0
	* Stream intersect mode
	* MI (LAMI): versioning, progress status, more sections
	* Move LAMI specification to another repository
	* Improved testing (unit and integration)
	* Code cleanup and testing of common utils functions
	* Various bugfixes

2016-03-07 LTTng analyses 0.4.3
	* Tests fixes (timezone issues)

2016-03-01 LTTng analyses 0.4.2
	* Packaging fixes

2016-02-29 LTTng analyses 0.4.1
	* Packaging fixes

2016-02-26 LTTng analyses 0.4.0
	* Scheduler latency analyses
	* Priority fields in CPU and latency analyses
	* Machine Interface (json) output
	* Period-based analyses (begin and end events)
	* Refactoring/Bugfixes/Cleanup
	* Basic testing infrastructure

2015-07-13 LTTng analyses 0.3.0
	* Merge pull request #23 from mjeanson/master
	* Convert README to reStructuredText
	* Fix pep8 errors
	* Refactor in a single package with subpackages
	* fix: stats with 0 requests
	* Check for babeltrace python package on installation
	* Define version once per package only
	* Add ChangeLog file

2015-04-20 LTTng analyses 0.2.0
	* Merge pull request #22 from abusque/refactor-syscallstats
	* Bump version to 0.2
	* Refactor syscallstats script to use new analysis backend
	* Rename min/max attributes to avoid collision with built-ins
	* Merge pull request #21 from abusque/decouple-io
	* Implement check for --end argument before start of trace
	* Style: fix indentation in _get_io_requests
	* Fix: set pid correctly on FileStats init
	* Fix typo in _fix_context_pid
	* Fix: use TID instead of PID in file stats if PID is None
	* Refactor io latency freq output
	* Lint: remove unused import, fix 'dangerous' default args
	* Refactor io top and log views
	* Remove deprecated --extra argument
	* Fix: correct typo and existence test in fd getter
	* Fix: correct typo in ns_to_hour_nsec output
	* Style: fix pylint/pep8 style issues
	* Replace map() by list comprehension in disk latency stats
	* Refactor IO Latency stats output methods
	* Add generators to iterate over io requests
	* Add method to compare equivalent io operations
	* Fix: properly handle empty filters for IO file stats
	* Fix FileStats reset() function
	* Move _filter_process method to base command class
	* Make _arg_pid_list list of ints instead of strings
	* Refactor iotop per file analysis and output
	* Refactor iotop output methods
	* Add _print_ascii_graph method to simplify output of graphs
	* Rename filter predicates to indicate visibility
	* Remove deprecated breakcb in IO command
	* Remove unused _compute_stats method from commands
	* Rename IO command for consistency with other commands
	* Track FDs chronologically in IO analysis
	* Add timestamp to create/close FD notifications
	* Remove dead code from IO cli
	* Reset FD in IO Analysis
	* Add support for pwrite* and pread* I/O syscalls
	* Implement syscall I/O analysis
	* Move returned_size attribute from SyscallIORequest into ReadWriteIORequest
	* Send create process and fd notification on statedump events
	* Send fd create and close notifications on sched events
	* Fix: send create_fd notification for open io requests
	* Add OP_READ_WRITE IO operation type for syscalls which both read and write
	* Use a single method to track io request exits
	* Refactor/rewrite IO state provider
	* Refactor syscall analysis to use new SyscallEvent class
	* Refactor NetStateProvider to use new SyscallEvent and io rq objects
	* Refactor MemStateProvider to use new SyscallEvent and io rq objects
	* Remove pending_syscalls array from State class
	* Refactor statedump provider to track only state and not analysis related attributes
	* Don't set deprecated parent_pid on FD object
	* Use SyscallEvent objects in syscall state provider
	* Remove Syscalls_stats class
	* Remove analysis related attributes from FD class, add factory to create from open rq
	* Add get_fd_type method to retrieve fd type from syscall name
	* Add more IORequest classes, and io_rq attr to SyscallEvent
	* Set SyscallEvent name using get_syscall_name method
	* Remove analysis related attributes from Process state class
	* Add more dup open syscalls, remove generic filenames from SyscallConsts
	* Fix get_syscall_name string indexing
	* Move IO syscalls handling into separate provider
	* Strip prefixes from syscall names for brevity
	* Merge branch 'master' into decouple-io
	* Merge pull request #20 from abusque/linting
	* Rename state to _state in providers for consistency
	* Rename irq start/stop timestamps to begin/end for consistency
	* Refactor IO Requests mechanism and (block I/O) analysis
	* Track network usage in IO analysis
	* Separate syscalls and io analyses
	* Use del instead of pop when possible with fds and remove unused attributes
	* Move date args processing to command, more linting
	* Linting: rename p* to pattern
	* Linting of common.py and related code
	* Fix: make the regex strings raw strings
	* fix for unknown pid in io.py
	* Fix syscallstats command description method names
	* Add IO analysis separate from syscalls
	* Merge pull request #19 from jdesfossez/dev
	* Fix: process the sched_switch for the swapper
	* Fix: handle the case of missing PID
	* Merge pull request #18 from abusque/decouple-cputop
	* Revert accidental partial commit of syscalls.py

irq.py
#!/usr/bin/env python3
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in
# all copies or substantial portions of the Software.

import argparse
import sys
import statistics

try:
    from babeltrace import TraceCollection
except ImportError:
    # quick fix for debian-based distros
    sys.path.append("/usr/local/lib/python%d.%d/site-packages" %
                    (sys.version_info.major, sys.version_info.minor))
    from babeltrace import TraceCollection
from LTTngAnalyzes.common import NSEC_PER_SEC, ns_to_asctime, IRQ, \
    ns_to_hour_nsec, process_date_args
from LTTngAnalyzes.progressbar import progressbar_setup, progressbar_update, \
    progressbar_finish
from LTTngAnalyzes.state import State
from ascii_graph import Pyasciigraph


class IrqStats():
    def __init__(self, traces):
        self.trace_start_ts = 0
        self.trace_end_ts = 0
        self.traces = traces
        self.state = State()

    def run(self, args):
        """Process the trace"""
        self.current_sec = 0
        self.start_ns = 0
        self.end_ns = 0

        progressbar_setup(self, args)
        if not args.begin:
            started = 1
        else:
            started = 0
        for event in self.traces.events:
            progressbar_update(self, args)
            if args.begin and started == 0 and event.timestamp >= args.begin:
                started = 1
                self.trace_start_ts = event.timestamp
                self.reset_total(event.timestamp)
            if args.end and event.timestamp > args.end:
                break
            if self.start_ns == 0:
                self.start_ns = event.timestamp
            if self.trace_start_ts == 0:
                self.trace_start_ts = event.timestamp
            self.end_ns = event.timestamp
            self.check_refresh(args, event)
            self.trace_end_ts = event.timestamp

            if event.name == "sched_switch":
                self.state.sched.switch(event)
            elif event.name == "irq_handler_entry":
                self.state.irq.hard_entry(event)
            elif event.name == "irq_handler_exit":
                self.state.irq.hard_exit(event, args)
            elif event.name == "softirq_entry":
                self.state.irq.soft_entry(event)
            elif event.name == "softirq_exit":
                self.state.irq.soft_exit(event, args)
            elif event.name == "softirq_raise":
                self.state.irq.soft_raise(event)
        if args.refresh == 0:
            # stats for the whole trace
            self.output(args, self.trace_start_ts, self.trace_end_ts, final=1)
        else:
            # stats only for the last segment
            self.output(args, self.start_ns, self.trace_end_ts, final=1)
        progressbar_finish(self, args)

    def check_refresh(self, args, event):
        """Check if we need to output something"""
        if args.refresh == 0:
            return
        event_sec = event.timestamp / NSEC_PER_SEC
        if self.current_sec == 0:
            self.current_sec = event_sec
        elif self.current_sec != event_sec and \
                (self.current_sec + args.refresh) <= event_sec:
            self.output(args, self.start_ns, event.timestamp)
            self.reset_total(event.timestamp)
            self.current_sec = event_sec
            self.start_ns = event.timestamp

    def compute_stdev(self, irq):
        values = []
        raise_delays = []
        stdev = {}
        for j in irq["list"]:
            delay = j.stop_ts - j.start_ts
            values.append(delay)
            if j.raise_ts == -1:
                continue
            # Raise latency (only for some softirqs)
            r_d = j.start_ts - j.raise_ts
            raise_delays.append(r_d)
        if irq["count"] < 2:
            stdev["duration"] = "?"
        else:
            stdev["duration"] = "%0.03f" % (statistics.stdev(values) / 1000)
        # format string for the raise if present
        if irq["raise_count"] >= 2:
            stdev["raise"] = "%0.03f" % (statistics.stdev(raise_delays)/1000)
        return stdev

    def irq_list_to_freq(self, irq, _min, _max, res):
        step = (_max - _min) / res
        if step == 0:
            return
        buckets = []
        values = []
        graph = Pyasciigraph()
        for i in range(res):
            buckets.append(i * step)
            values.append(0)
        for i in irq["list"]:
            v = (i.stop_ts - i.start_ts) / 1000
            b = min(int((v-_min)/step), res - 1)
            values[b] += 1
        g = []
        i = 0
        for v in values:
            g.append(("%0.03f" % (i * step + _min), v))
            i += 1
        for line in graph.graph('Frequency distribution', g, info_before=True):
            print(line)
        print("")

    # FIXME: there must be a way to make that more complicated/ugly
    def filter_irq(self, args, i):
        if i.irqclass == IRQ.HARD_IRQ:
            if args.irq_filter_list is not None:
                if len(args.irq_filter_list) > 0 and \
                        str(i.nr) not in args.irq_filter_list:
                    return False
                else:
                    return True
            if args.softirq_filter_list is not None and \
                    len(args.softirq_filter_list) > 0:
                return False
        else:
            if args.softirq_filter_list is not None:
                if len(args.softirq_filter_list) > 0 and \
                        str(i.nr) not in args.softirq_filter_list:
                    return False
                else:
                    return True
            if args.irq_filter_list is not None and \
                    len(args.irq_filter_list) > 0:
                return False
        raise Exception("WTF")

    def log_irq(self, args):
        fmt = "[{:<18}, {:<18}] {:>15} {:>4}  {:<9} {:>4}  {:<22}"
        title_fmt = "{:<20} {:<19} {:>15} {:>4}  {:<9} {:>4}  {:<22}"
        print(title_fmt.format("Begin", "End", "Duration (us)", "CPU",
                               "Type", "#", "Name"))
        for i in self.state.interrupts["irq-list"]:
            if not self.filter_irq(args, i):
                continue
            if i.irqclass == IRQ.HARD_IRQ:
                name = self.state.interrupts["names"][i.nr]
                irqtype = "IRQ"
            else:
                name = IRQ.soft_names[i.nr]
                irqtype = "SoftIRQ"
            if i.raise_ts != -1:
                raise_ts = " (raised at %s)" % \
                           (ns_to_hour_nsec(i.raise_ts, args.multi_day,
                                            args.gmt))
            else:
                raise_ts = ""
            print(fmt.format(ns_to_hour_nsec(i.start_ts, args.multi_day,
                                             args.gmt),
                             ns_to_hour_nsec(i.stop_ts, args.multi_day,
                                             args.gmt),
                             "%0.03f" % ((i.stop_ts - i.start_ts) / 1000),
                             "%d" % i.cpu_id, irqtype, i.nr, name + raise_ts))

    def print_irq_stats(self, args, dic, name_table, filter_list, header):
        header_output = 0
        for i in sorted(dic.keys()):
            if len(filter_list) > 0 and str(i) not in filter_list:
                continue
            name = name_table[i]
            stdev = self.compute_stdev(dic[i])

            # format string for the raise if present
            if dic[i]["raise_count"] < 2:
                raise_stats = " |"
            else:
                r_avg = dic[i]["raise_total"] / (dic[i]["raise_count"] * 1000)
                raise_stats = " | {:>6} {:>12} {:>12} {:>12} {:>12}".format(
                    dic[i]["raise_count"],
                    "%0.03f" % (dic[i]["raise_min"] / 1000),
                    "%0.03f" % r_avg,
                    "%0.03f" % (dic[i]["raise_max"] / 1000),
                    stdev["raise"])

            # final output
            if dic[i]["count"] == 0:
                continue
            avg = "%0.03f" % (dic[i]["total"] / (dic[i]["count"] * 1000))
            format_str = '{:<3} {:<18} {:>5} {:>12} {:>12} {:>12} ' \
                         '{:>12} {:<60}'
            s = format_str.format("%d:" % i, "<%s>" % name, dic[i]["count"],
                                  "%0.03f" % (dic[i]["min"] / 1000),
                                  "%s" % (avg),
                                  "%0.03f" % (dic[i]["max"] / 1000),
                                  "%s" % (stdev["duration"]),
                                  raise_stats)
            if args.freq or header_output == 0:
                print(header)
                header_output = 1
            print(s)
            if args.freq:
                self.irq_list_to_freq(dic[i], dic[i]["min"] / 1000,
                                      dic[i]["max"] / 1000,
                                      args.freq_resolution)

    def output(self, args, begin_ns, end_ns, final=0):
        if args.no_progress:
            clear_screen = ""
        else:
            clear_screen = "\r" + self.pbar.term_width * " " + "\r"
        date = '%s to %s' % (ns_to_asctime(begin_ns), ns_to_asctime(end_ns))
        print(clear_screen + date)
        if args.irq_filter_list is not None:
            header = ""
            header += '{:<52} {:<12}\n'.format("Hard IRQ", "Duration (us)")
            header += '{:<22} {:<14} {:<12} {:<12} {:<10} ' \
                      '{:<12}\n'.format("", "count", "min", "avg", "max",
                                        "stdev")
            header += ('-'*82 + "|")
            self.print_irq_stats(args, self.state.interrupts["hard-irqs"],
                                 self.state.interrupts["names"],
                                 args.irq_filter_list, header)
            print("")

        if args.softirq_filter_list is not None:
            header = ""
            header += '{:<52} {:<52} {:<12}\n'.format("Soft IRQ",
                                                      "Duration (us)",
                                                      "Raise latency (us)")
            header += '{:<22} {:<14} {:<12} {:<12} {:<10} {:<4} {:<3} {:<14} '\
                      '{:<12} {:<12} {:<10} ' \
                      '{:<12}\n'.format("", "count", "min", "avg", "max",
                                        "stdev", " |", "count", "min",
                                        "avg", "max", "stdev")
            header += '-' * 82 + "|" + '-' * 60
            self.print_irq_stats(args, self.state.interrupts["soft-irqs"],
                                 IRQ.soft_names, args.softirq_filter_list,
                                 header)
            print("")

        if args.log:
            self.log_irq(args)

    def reset_total(self, start_ts):
        self.state.interrupts["hard_count"] = 0
        self.state.interrupts["soft_count"] = 0
        self.state.interrupts["irq-list"] = []
        for i in self.state.interrupts["hard-irqs"].keys():
            self.state.interrupts["hard-irqs"][i] = self.state.irq.init_irq()
        for i in self.state.interrupts["soft-irqs"].keys():
            self.state.interrupts["soft-irqs"][i] = self.state.irq.init_irq()

if __name__ == "__main__":
    parser = argparse.ArgumentParser(description='Memory usage analysis')
    parser.add_argument('path', metavar="<path/to/trace>", help='Trace path')
    parser.add_argument('-r', '--refresh', type=int,
                        help='Refresh period in seconds', default=0)
    parser.add_argument('--gmt', action="store_true",
                        help='Manipulate timestamps based on GMT instead '
                        'of local time')
    parser.add_argument('--top', type=int, default=10,
                        help='Limit to top X TIDs (default = 10)')
    parser.add_argument('--no-progress', action="store_true",
                        help='Don\'t display the progress bar')
    parser.add_argument('--details', action="store_true",
                        help='Display all the IRQs details')
    parser.add_argument('--thresh', type=int, default=0,
                        help='Threshold in ns for the detailled view')
    parser.add_argument('--irq', type=str, default=0,
                        help='Show results only for the list of IRQ')
    parser.add_argument('--softirq', type=str, default=0,
                        help='Show results only for the list of SoftIRQ')
    parser.add_argument('--freq', action="store_true",
                        help='Show the frequency distribution of handler '
                             'duration')
    parser.add_argument('--freq-resolution', type=int, default=20,
                        help='Frequency distribution resolution (default 20)')
    parser.add_argument('--max', type=float, default=-1,
                        help='Filter out, duration longer than max usec')
    parser.add_argument('--min', type=float, default=-1,
                        help='Filter out, duration shorter than min usec')
    parser.add_argument('--log', action="store_true",
                        help='Display the interrupt in the order they were '
                             'handled')
    parser.add_argument('--begin', type=str, help='start time: '
                                                  'hh:mm:ss[.nnnnnnnnn]')
    parser.add_argument('--end', type=str, help='end time: '
                                                'hh:mm:ss[.nnnnnnnnn]')
    parser.add_argument('--timerange', type=str, help='time range: '
                                                      '[begin,end]')
    args = parser.parse_args()

    args.irq_filter_list = None
    args.softirq_filter_list = None

    if args.irq:
        args.irq_filter_list = args.irq.split(",")

    if args.softirq:
        args.softirq_filter_list = args.softirq.split(",")
    if args.irq_filter_list is None and args.softirq_filter_list is None:
        args.irq_filter_list = []
        args.softirq_filter_list = []
    if args.max == -1:
        args.max = None
    if args.min == -1:
        args.min = None

    traces = TraceCollection()
    handle = traces.add_traces_recursive(args.path, "ctf")
    if handle is None:
        sys.exit(1)

    process_date_args(args, handle)

    c = IrqStats(traces)

    c.run(args)

    for h in handle.values():
        traces.remove_trace(h)
